src_c/camera_windows.c

```diff
@@ -0,0 +1,859 @@
+     * Sadly, using the source reader in this way would only let me open
+     * the webcam in a single size, so I replaced it with a video processor
+     * MFT. (Which also helps with the flip controls). */
+    /*
+    IMFAttributes *rsa;
+    MFCreateAttributes(&rsa, 1);
+    rsa->lpVtbl->SetUINT32(rsa, &MF_SOURCE_READER_ENABLE_VIDEO_PROCESSING, 1);
+
+    hr = MFCreateSourceReaderFromMediaSource(source, rsa, &reader);
+    RELEASE(rsa);
+    CHECKHR(hr);
+    */
+
+    hr = MFCreateSourceReaderFromMediaSource(source, NULL, &reader);
+    self->reader = reader;
+    RELEASE(source);
+    HANDLEHR(hr);
+
+    if(!_select_source_type(self, &source_type)) {
+        return 0;
+    }
+
+    hr = reader->lpVtbl->SetCurrentMediaType(reader, FIRST_VIDEO, NULL,
+                                             source_type);
+    HANDLEHR(hr);
+
+    hr = CoCreateInstance(&CLSID_VideoProcessorMFT, NULL,
+                          CLSCTX_INPROC_SERVER, &IID_IMFTransform,
+                          &transform);
+    self->transform = transform;
+    HANDLEHR(hr);
+
+    hr = transform->lpVtbl->QueryInterface(transform,
+                                           &IID_IMFVideoProcessorControl,
+                                           &control);
+    self->control = control;
+    HANDLEHR(hr);
+
+    hr = transform->lpVtbl->SetInputType(transform, 0, source_type, 0);
+    HANDLEHR(hr);
+
+    if (!_select_conversion_type(self, &conv_type)) {
+        return 0;
+    }
+
+    hr = transform->lpVtbl->SetOutputType(transform, 0, conv_type, 0);
+    HANDLEHR(hr);
+
+    hr = self->transform->lpVtbl->GetOutputStreamInfo(self->transform, 0,
+                                                      &info);
+    HANDLEHR(hr);
+
+    hr = MFCreateMemoryBuffer(info.cbSize, &self->buf);
+    HANDLEHR(hr);
+
+    self->t_handle = CreateThread(NULL, 0, update_function, self, 0, NULL);
+
+    self->open = 1; /* set here, since this shouldn't happen on error */
+    self->t_error = S_OK;
+    self->t_error_line = 0;
+    return 1;
+
+    cleanup:
+        windows_close_device(self);
+        return 0;
+}
+
+int
+windows_close_device(pgCameraObject *self)
+{
+    self->open = 0;
+    if (self->t_handle) {
+        WaitForSingleObject(self->t_handle, 3000);
+    }
+
+    RELEASE(self->reader);
+    RELEASE(self->transform);
+    RELEASE(self->control);
+    RELEASE(self->buf);
+    RELEASE(self->raw_buf);
+
+    return 1;
+}
+
+int
+windows_init_device(pgCameraObject* self)
+{
+    HRESULT hr;
+
+    /* I wanted to use COINIT_MULTITHREADED, but something inside SDL started
+     * by display.set_mode() uses COINIT_APARTMENTTHREADED, and you can't have
+     * a thread in two modes at once */
+    hr = CoInitializeEx(0, COINIT_APARTMENTTHREADED);
+    CHECKHR(hr);
+
+    hr = MFStartup(MF_VERSION, MFSTARTUP_LITE);
+    CHECKHR(hr);
+}
+
+int
+windows_dealloc_device(pgCameraObject *self)
+{
+    PyMem_Free(self->device_name);
+
+    MFShutdown();
```

Comment:
I guess it could be useful to check the error code here for printing debugging:

https://docs.microsoft.com/en-us/windows/win32/api/mfapi/nf-mfapi-mfshutdown#return-value